examples/TestEm3/electrons.cu

```diff
@@ -258,11 +265,83 @@ void TransportElectrons(Track *electrons, const adept::MParray *active,
+  }
+}
+
+
+// Instantiate kernels for electrons and positrons.
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void TransportElectrons(Track *electrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  TransportElectrons</*IsElectron*/ true>(electrons, active, secondaries, activeQueue, globalScoring, scoringPerVolume);
+}
+__global__ __launch_bounds__(ThreadsPerBlock, MinBlocksPerSM)
+void TransportPositrons(Track *positrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  TransportElectrons</*IsElectron*/ false>(positrons, active, secondaries, activeQueue, globalScoring,
+                                           scoringPerVolume);
+}
+
+template<bool IsElectron, int ProcessIndex>
+__global__
+void ComputeInteraction(Track *electrons, const adept::MParray *active, Secondaries secondaries,
+                        adept::MParray *activeQueue, GlobalScoring *globalScoring,
+                        ScoringPerVolume *scoringPerVolume)
+{
+  constexpr unsigned int sharedSize = 12250;
+  __shared__ int candidates[sharedSize];
+  __shared__ unsigned int counter;
+  __shared__ unsigned int noopCounter;
+  counter = 0;
+  noopCounter = 0;
+
+  for (int i = threadIdx.x; i < sharedSize; i += blockDim.x) {
+    candidates[i] = -1;
+  }
+  __syncthreads();
+
+  const int activeSize = active->size();
+  assert(activeSize < sizeof(g_nextInteractionForEl)/sizeof(g_nextInteractionForEl[0]));
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < activeSize; i += blockDim.x * gridDim.x) {
+    const auto winnerProcess = IsElectron ? g_nextInteractionForEl[i] : g_nextInteractionForPos[i];
+
+    if (winnerProcess == ProcessIndex) {
+      const auto destination = atomicInc(&counter, (unsigned int)-1);
+      candidates[destination % sharedSize] = i;
+    } else atomicInc(&noopCounter, (unsigned int)-1);
+  }
+
+  __syncthreads();
+  assert(counter < sharedSize);
+  if (threadIdx.x == 0 && counter >= sharedSize) {
+    printf("Error: Shared queue for %d exhausted to %d in %s:%d\n", ProcessIndex, counter, __FILE__, __LINE__);
+    asm("trap;");
+  }
+
+  for (int i = threadIdx.x; i < counter; i += blockDim.x) {
+    const int slot = (*active)[candidates[i]];
+    Track &currentTrack = electrons[slot];
+    auto volume         = currentTrack.navState.Top();
+    int volumeID        = volume->id();
+    int theMCIndex      = MCIndex[volumeID];
+
     const double energy   = currentTrack.energy;
     const double theElCut = g4HepEmData.fTheMatCutData->fMatCutData[theMCIndex].fSecElProdCutE;
 
-    switch (winnerProcessIndex) {
-    case 0: {
+    RanluxppDouble newRNG(currentTrack.rngState.BranchNoAdvance());
```

Comment:
This *must* use `Branch()` (with `Advance`ing) or `newRNG` has correlations to the primary's RNG state, see #129. In turn, the calls to `newRNG.Advance();` and `currentTrack.rngState.Advance();` in the `Transport` kernel above are *not* needed anymore.